### DISCUSSION TITLE
supercomputers deployment with unidist MPI backend

### DISCUSSION THREAD

**Discussion by luweizheng** [modin-project/modin#6073] [Q&A] [answered]
hi all!

I read modin's doc and do not find any example on how to start `unidist` backend on supercomputers.
Supercomputers have job schedulers like SLURM or PBS. I can start modin on ray following ray's SLURM deployment [document](https://docs.ray.io/en/latest/cluster/vms/user-guides/community/slurm.html?highlight=slurm#python-interface-slurm-scripts). What about `unidist`? I see there is MPI backend in `unidist`. And supercomputers usually have InfiniBand and MPI support.

Now I am using Intel MPI which can support spawning MPI processes without using `mpiexec` command.

What I do:

```
### SLURM script
export MODIN_ENGINE=unidist
export UNIDIST_BACKEND=mpi

# export MPI hosts here 
# as docs say that we need to set hosts: https://unidist.readthedocs.io/en/latest/using_unidist/unidist_on_mpi.html#id1
export UNIDIST_MPI_HOSTS=$joined


python modin_query.py
```

Now I get the following error message:

```
[mpiexec@cpu64c-5] check_exit_codes (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:117): unable to run bstrap_proxy on cpu64c-5 (pid 28288, exit code 256)
[mpiexec@cpu64c-5] poll_for_event (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:159): check exit codes error
[mpiexec@cpu64c-5] HYD_dmx_poll_wait_for_proxy_event (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:212): poll for event error
[mpiexec@cpu64c-5] HYD_bstrap_setup (../../../../../src/pm/i_hydra/libhydra/bstrap/src/intel/i_hydra_bstrap.c:1065): error waiting for event
[mpiexec@cpu64c-5] HYD_print_bstrap_setup_error_message (../../../../../src/pm/i_hydra/mpiexec/intel/i_mpiexec.c:1027): error setting up the bootstrap proxies
[mpiexec@cpu64c-5] Possible reasons:
[mpiexec@cpu64c-5] 1. Host is unavailable. Please check that all hosts are available.
[mpiexec@cpu64c-5] 2. Cannot launch hydra_bstrap_proxy or it crashed on one of the hosts. Make sure hydra_bstrap_proxy is available on all hosts and it has right permissions.
[mpiexec@cpu64c-5] 3. Firewall refused connection. Check that enough ports are allowed in the firewall and specify them with the I_MPI_PORT_RANGE variable.
[mpiexec@cpu64c-5] 4. slurm bootstrap cannot launch processes on remote host. You may try using -bootstrap option to select alternative launcher.
```

It will be helpful if there are official `unidist` docs so I can know if this is the problem with my configuration or if the way I use `unidist` is wrong.

**Comment by mvashishtha** [comment]
@luweizheng thank you for creating this issue. Unfortunately I know very little about MPI and unidist.

Is the documentation [here](https://unidist.readthedocs.io/en/latest/using_unidist/unidist_on_mpi.html) helpful at all? The error messages seem to say that unidist was unable to set up the remote workers. Would it help to customize the worker hosts [here](https://unidist.readthedocs.io/en/latest/using_unidist/unidist_on_mpi.html#id1)? Have you checked whether your controller node is able to communicate with the workers?

cc'ing @YarShev @modin-project/unidist-core for help. I think they can be more helpful.

**Reply by luweizheng** [reply]
Thanks for your reply. Both of the settings from unidist docs are already configured in my case. 
Hope developers from unidist could help.

**Reply by YarShev** [reply]
Hi @luweizheng, thank you for this post and sorry for the delayed response. I am currently on a trip and did not have a chance to reply here earlier. So far, we do not have any page in unidist docs yet on how to launch unidist on MPI using SLURM, but what I see looking at your steps with unidist configuration settings seems correct to me. In addition, I recommend you to set `UNIDIST_CPUS` environment variable in order to spawn the needed number of MPI workers in your cluster. `UNIDIST_CPUS` should be equal to the `number of processes + 1` you want to allocate with SLURM because one of MPI workers is going to be used for some monitoring tasks in unidist. That said, I think the issue you are experiencing comes down to the configuration settings of a concrete MPI implementation, namely, Intel MPI in your case. Could you take a look into this [post](https://community.intel.com/t5/Intel-oneAPI-HPC-Toolkit/Intel-MPI-Unable-to-run-bstrap-proxy-error-setting-up-the/m-p/1204677), pick the suggestions/solutions up from it and try those out? Please let us know if that helps.

**Reply by luweizheng** [reply]
So currently modin can only be hosted on ray? I must connect a ray cluster and submit my job to this ray cluster?

**Reply by YarShev** [reply]
To run Modin with unidist on MPI you don't need Ray at all. We haven't tested Modin with unidist on MPI thoroughly in a cluster environment yet, but, theoretically, it should work. In the example posted here you use unidist on MPI with the dynamic spawning of workers. Though it works well in a single node, I am not sure the dynamic spawning would work with a SLURM environment. I have to ask developers of the Intel MPI implementation on this matter. Meanwhile, I think the most appropriate way to run Modin with unidist on MPI using SLURM is to run the MPI backend in SPMD mode as shown [here](https://unidist.readthedocs.io/en/latest/using_unidist/unidist_on_mpi.html#id2). You should initialize unidist at the beginning of your script so after that you would get Controller/Worker mode (main process you interact with and worker processes performing tasks).

```bash
### SLURM script
export MODIN_ENGINE=unidist
export UNIDIST_BACKEND=mpi
export UNIDIST_IS_MPI_SPAWN_WORKERS=False
mpiexec -host host1:n1,...,hostM:nM python script.py
```

```python
### Python script
# initialize unidist at the beginning
import unidist
unidist.init()
```

For more details about Intel MPI usage with SLURM please look into [here](https://slurm.schedmd.com/mpi_guide.html#intel_mpi).

**Comment by no-ponomarev** [comment]
@luweizheng could you, please, provide more details about you use-case. You are starting one process application and then spawn the processes with Intel MPI? What `slurm` features are you using exactly?

**Reply by luweizheng** [reply, marked as answer]
Now I can run on unidist MPI backend successfully!

My SLURM script:

```
export MODIN_ENGINE=unidist
export UNIDIST_BACKEND=mpi

python xxx.py
```

remove all `ray` code and add `import unidist;unidist.init()l` into `xxx.py`:

```
import unidist
unidist.init()
```

I use Intel MPI and Intel MPI can automatically detect hosts by reading the `$SLURM_JOB_NODELIST` from SLURM.

Maybe we should update the docs? I can help to contribute. Where should I put these things into the docs?

**Reply by YarShev** [reply]
Happy to hear that you were able to run unidist on MPI. It would be great if you could help us update the docs on this matter. Could you please update this page ([link](https://unidist.readthedocs.io/en/latest/using_unidist/unidist_on_mpi.html)) by adding an additional section named like `Deploying on Slurm`.

Meanwhile, I would like to ask you a few questions to make sure you got the expected behavior.

1. Did you set `UNIDIST_MPI_HOSTS`? If not, are you sure that unidist on MPI spawned workers on all nodes listed in `SLURM_JOB_NODELIST`?
2. Did you set `UNIDIST_CPUS` to a specific value? If not, did you have the expected number of worker processes on all nodes listed in `SLURM_JOB_NODELIST`?
3. Which SLURM commands did you use to allocate the necessary amount of workers per node? Did you see that each node was able to run the requested amount of workers?

**Reply by luweizheng** [reply]
> 1. Did you set `UNIDIST_MPI_HOSTS`? If not, are you sure that unidist on MPI spawned workers on all nodes listed in `SLURM_JOB_NODELIST`?

No. If I set `UNIDIST_MPI_HOSTS=cpu24c-15,cpu24c-16,cpu24c-17,cpu24c-18`, where cpu24c-15 is the hostname on my site, I will get the following error:

```
[proxy:1:2@cpu24c-17] main (../../../../../src/pm/i_hydra/proxy/proxy.c:938): assert (proxy_params.immediate.process.num_children * sizeof(int)) failed
[proxy:1:3@cpu24c-18] main (../../../../../src/pm/i_hydra/proxy/proxy.c:938): assert (proxy_params.immediate.process.num_children * sizeof(int)) failed
[proxy:1:1@cpu24c-16] main (../../../../../src/pm/i_hydra/proxy/proxy.c:938): assert (proxy_params.immediate.process.num_children * sizeof(int)) failed
srun: error: cpu24c-17: task 2: Exited with exit code 5
srun: error: cpu24c-16: task 1: Exited with exit code 5
srun: error: cpu24c-18: task 3: Exited with exit code 5
```

> 2. Did you set `UNIDIST_CPUS` to a specific value? If not, did you have the expected number of worker processes on all nodes listed in `SLURM_JOB_NODELIST`?

`UNIDIST_CPUS` does not have any effect on my environment. This workload takes about 40 to 50 seconds. I am not sure if `UNIDIST_CPUS` can affect performance.

> 3. Which SLURM commands did you use to allocate the necessary amount of workers per node? Did you see that each node was able to run the requested amount of workers?

Here is my SLURM options to allocate resources.
```
#SBATCH --nodes=4
#SBATCH --cpus-per-task=24
#SBATCH --ntasks-per-node=1
```

Each node will start multiple workers, I use `top` and `htop` to watch the CPU workloads. 

Case 1: 

```
#SBATCH --cpus-per-task=24
#SBATCH --ntasks-per-node=1
```
It seems that the workload cannot run on all CPU cores even I set `UNIDIST_CPUS` to a very large value like 48. In my case, 6 python processes on a 24 cores machine. Some cores' utilization is 0.0%. Some cores' utilization is 100%.


Case 2:

```
#SBATCH --cpus-per-task=4
#SBATCH --ntasks-per-node=6
```

More cores are used.


Case 3:

```
#SBATCH --cpus-per-task=1
#SBATCH --ntasks-per-node=24
```

Every core's utilization is 100%. But the overall speed is very slow.

**Reply by YarShev** [reply]
> No. If I set UNIDIST_MPI_HOSTS=cpu24c-15,cpu24c-16,cpu24c-17,cpu24c-18, where cpu24c-15 is the hostname on my site, I will get the following error:

If you don't set `UNIDIST_MPI_HOSTS`, I am not sure that MPI processes are spawned across all nodes you listed in `SLURM_JOB_NODELIST`. Do you really see MPI processes spawned on all nodes without setting `UNIDIST_MPI_HOSTS`?

> UNIDIST_CPUS does not have any effect on my environment. This workload takes about 40 to 50 seconds. I am not sure if UNIDIST_CPUS can affect performance.

`UNIDIST_CPUS` should affect performance because it specifies how many MPI workers should be spawned.

> Here is my SLURM options to allocate resources.

Regarding your 3 cases, the third one is the best. NTasks per node can vary but cpus per task should be equal to 1. This is how Modin on Ray/Dask works.

Anyway, I really encourage you to try to run unidist on MPI in SPMD mode as I said above because there is a direct mapping between SLURM commands and mpiexec. When you run a program with mpiexec, it respects/copies settings from the SLURM batch job.

Look forward to your response.

**Reply by luweizheng** [reply]
> If you don't set UNIDIST_MPI_HOSTS, I am not sure that MPI processes are spawned across all nodes you listed in SLURM_JOB_NODELIST. Do you really see MPI processes spawned on all nodes without setting UNIDIST_MPI_HOSTS?

Here are two screen snapshots on 2 nodes with slurm setting 1. MPI processes spawn on multiple nodes.

<img width="1220" alt="image" src="https://github.com/modin-project/modin/assets/57242864/fe61e6e8-7bac-4806-b15e-dd2381ac467e">

<img width="1220" alt="image" src="https://github.com/modin-project/modin/assets/57242864/00e06d89-73a1-42d8-9c44-e5baea813866">

**Reply by YarShev** [reply]
@luweizheng, I see, thanks for the screenshots. Apperently, `MPI.Comm.Spawn` somehow detects the settings from `SLURM_JOB_NODELIST`. Also, it is suprising to me that with the default value of `UNIDIST_CPUS` you get the number of spawned MPI processes equal to the number of CPUs on each node. We should talk to developers of Intel MPI and investigate the behavior of `MPI.Comm.Spawn` more thoroughly. I would really appreciate if you could share your SLURM script and the workload itself in order for me to get more insight regarding the behavior. If you feel that these things can't be shared publicly, please reach me out directly via email iaroslav.igoshev@intel.com.

**Reply by YarShev** [reply]
With respect to performance, we are currently working hard to get things work faster. Hopefully, unidist 0.4.0 will contain all the performance improvements. I was wondering if you could tell us how much faster Ray is than unidist on MPI on your workload?

**Reply by luweizheng** [reply]
I can share the workloads with you. Actually, I am doing the tpch benchmarks. Let me do some refactoring and I will give the details on how to deploy on unidst.
What's more, when running the tpch benchmarks. It seems that some semantics of the MPI backend differ from ray. These differences cause exceptions when executing the same query on undist than ray. Maybe later I should open an issue on unidist.

**Reply by YarShev** [reply]
> I can share the workloads with you. Actually, I am doing the tpch benchmarks. Let me do some refactoring and I will give the details on how to deploy on unidst.

Sure, take your time.

> What's more, when running the tpch benchmarks. It seems that some semantics of the MPI backend differ from ray. These differences cause exceptions when executing the same query on undist than ray. Maybe later I should open an issue on unidist.

That is really interesting. It would be helpful for us if you could open an issue in unidist.

**Reply by luweizheng** [reply]
Hi @YarShev 

Here is my codebase:  https://github.com/dbiir/df-bench 

I put tpch benchmark in this repo.

To run the tpch benchmark, you should first generate data. You can refer to this [readme file](https://github.com/dbiir/df-bench/blob/main/tpch/README.md).

Now all queries in `pandas` and `polars` wok well. modin on ray and modin on unidist can run in my environment, too.
To run the unidist mpi backend on a supercomputing cluster. Please refer to scripts and python code in `modin` folder. There are README.md, .py files, and SLURM scripts.

If you have any problems running this program, please let me know.

Last time I mentioned that ray and unidist MPI are different. I think I was wrong. It seems that both ray and unidist fail on query 4. It seems that the problem is not related to unidist. Maybe because numpy is incompatible with modin or there is a bug in modin. My modin version is 0.20.1.

**Reply by YarShev** [reply]
I took a look at `slurm_modin_unidist_mpi.sh` script and it is a bit incorrect in terms of running Modin on unidist with MPI. Let me explain how the script runs Modin on unidist with MPI in your scenario. You use `mpiexec` to run `modin_unidist_query.py`, thus, multiple copies of `modin_unidist_query.py` will be launched on your nodes. The number of the application's instances depends on `--nodes=<N1>`, `--ntasks=<N2>`, `--ntasks-per-node=<N3>`. In your case the following happens.

1) Multiples copies of the application are launched on every node.
2) Each single copy spawns other workers because you don't set `UNIDIST_IS_MPI_SPAWN_WORKERS=False`.

This is not the right behavior and I am not sure if this works correctly on your side. I made a few fixes to the script (see below). Please launch the updated script and let me know if it works. The updated script should launch 96 copies of the application on 4 nodes, i.e., 24 copies on each node. This way of running a program is named as SPMD. Once you initialize unidist with `unidist.init()` in your application, the flow turns into Controller/Worker mode, i.e., you will have a master/controler process you interact with and multiple worker processes that execute the tasks.

```bash
$ git diff

diff --git a/tpch/modin/slurm_modin_unidist_mpi.sh b/tpch/modin/slurm_modin_unidist_mpi.sh
index 7eda347..b6df9f7 100644
--- a/tpch/modin/slurm_modin_unidist_mpi.sh
+++ b/tpch/modin/slurm_modin_unidist_mpi.sh
@@ -2,8 +2,9 @@

 #SBATCH --job-name=modin_unidist_mpi
 #SBATCH --nodes=4
-#SBATCH --cpus-per-task=24
-#SBATCH --ntasks-per-node=1
+#SBATCH --ntasks=96
+#SBATCH --ntasks-per-node=24
+#SBATCH --cpus-per-task=1
 #SBATCH --partition=cpu24c
 #SBATCH --output=modin_unidist_mpi.log

@@ -29,7 +30,7 @@ export UNIDIST_BACKEND=mpi
 # echo $UNIDIST_MPI_HOSTS
 # export UNIDIST_CPUS=24

-# export UNIDIST_IS_MPI_SPAWN_WORKERS=False
+export UNIDIST_IS_MPI_SPAWN_WORKERS=False
 export I_MPI_HYDRA_IFACE="ib0.8066"

 # change the `--path` to your local location
```

**Reply by luweizheng** [reply]
I change the slurm script as you suggested. 

`UNIDIST_IS_MPI_SPAWN_WORKERS ` affects performance.

There is a little problem here. I really appreciate it if you can check if anything wrong with my code.

In my code, I use `--queries` argument parser to parse different tpch queries. If I pass `1` to this option. I can get the result. Very Fast. However, if I change the `--queries` to `1 2 3`, it will be very slow. It just iterates with different queries and runs the specific query function.

Here is my code: https://github.com/dbiir/df-bench/blob/main/tpch/modin/modin_unidist_query.py

**Reply by YarShev** [reply]
> UNIDIST_IS_MPI_SPAWN_WORKERS  affects performance.

With this environment variable set up you should have the correct execution. If you don't set it, you will have a scenario I described above, which is wrong.

> There is a little problem here. I really appreciate it if you can check if anything wrong with my code.

> In my code, I use `--queries` argument parser to parse different tpch queries. If I pass `1` to this option. I can get the result. Very Fast. However, if I change the `--queries` to `1 2 3`, it will be very slow. It just iterates with different queries and runs the specific query function.

I took a look at the code and didn't see any issue related to the parsing of `--queries` option. It should be obvious that you have faster execution with `--queries 1` than with `--queries 1 2 3` because in the latter case you have much more operations to be executed with Modin. Maybe I am missing on something? What behavior do you expect to see?
